File: pat1089.java
/*
      思路：在归并排序过程中判断是否与排过序的序列相等
            如果存在相等序列则是归并排序，否则为插入排序
 */

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(reader.readLine());
        String[] line = reader.readLine().split(" ");
        int[] initSeq = new int[n];
        int[] partiallySortedSeq = new int[n];
        for (int i = 0; i < n; i++) {
            initSeq[i] = Integer.parseInt(line[i]);
        }
        line = reader.readLine().split(" ");
        for (int i = 0; i < n; i++) {
            partiallySortedSeq[i] = Integer.parseInt(line[i]);
        }
        Solution solution = new Solution(n, initSeq, partiallySortedSeq);
        if (solution.mergeSort()) {
            System.out.println("Merge Sort");
            solution.printArr(solution.initSeq);
        } else {
            System.out.println("Insertion Sort");
            solution.insertSort();
            solution.printArr(solution.partiallySortedSeq);
        }
    }

}

class Solution {
    int n;
    int[] initSeq;
    int[] partiallySortedSeq;

    public Solution(int n, int[] initSeq, int[] partiallySortedSeq) {
        this.n = n;
        this.initSeq = initSeq;
        this.partiallySortedSeq = partiallySortedSeq;
    }

    public boolean mergeSort() {
        boolean isMergeSort = false;
        for (int i = 2; i < 2 * n; i *= 2) {
            if (Arrays.equals(initSeq, partiallySortedSeq) && i != 2) {
                isMergeSort = true;
            }
            for (int j = 0; j < n; j += i) {
                Arrays.sort(initSeq, j, Math.min(n, j + i));
            }
            if (isMergeSort) {
                return true;
            }
        }
        return false;
    }

    public void insertSort() {
        // 一段序列有序，表示没有插入过数
        boolean hasInserted = false;
        for (int i = 1; i < n; i++) {
            int temp = partiallySortedSeq[i];
            int j = i;
            while (j > 0 && temp < partiallySortedSeq[j - 1]) {
                partiallySortedSeq[j] = partiallySortedSeq[j - 1];
                hasInserted = true;
                j--;
            }
            partiallySortedSeq[j] = temp;
            if (hasInserted) {
                break;
            }
        }
    }

    public void printArr(int[] arr) {
        System.out.print(arr[0]);
        for (int i = 1; i < arr.length; i++) {
            System.out.print(" " + arr[i]);
        }
    }

}

File: pat1089.cpp
/*
	思路：在归并排序过程中判断是否与排过序的序列相等
	      如果存在相等序列则是归并排序，否则为插入排序
*/

#include<iostream>
#include<algorithm>
using namespace std;

int n;
int init_seq[200];
int partially_sorted_seq[200];

bool is_arr_equal(int a[], int b[]) {
	for (int i = 0; i < n; i++) {
		if (a[i] != b[i]) {
			return false;
		}
	}
	return true;
}

bool merge_sort() {
	bool is_merge_sort = false;
	for (int i = 2; i < 2 * n; i *= 2) {
		if (is_arr_equal(init_seq, partially_sorted_seq) && i != 2) {
			is_merge_sort = true;
		}
		for (int j = 0; j < n; j += i) {
			sort(init_seq + j, init_seq + min(j + i, n));
		}
		if (is_merge_sort) {
			return true;
		}
	}

	return false;
}

void insert_sort() {
	bool has_inserted = false;	// 一段序列有序表示没有插入过数
	for (int i = 1; i < n; i++) {
		int temp = partially_sorted_seq[i];
		int j = i;
		while (j > 0 && temp < partially_sorted_seq[j - 1]) {
			partially_sorted_seq[j] = partially_sorted_seq[j - 1];
			j--;
			has_inserted = true;
		}
		partially_sorted_seq[j] = temp;
		if (has_inserted) {
			break;
		}
	}

}



void print_arr(int a[]) {
	cout << a[0];
	for (int i = 1; i < n; i++) {
		cout << " " << a[i];
	}
}

int main() {


	cin >> n;
	for (int i = 0; i < n; i++) {
		cin >> init_seq[i];
	}
	for (int i = 0; i < n; i++) {
		cin >> partially_sorted_seq[i];
	}


	if (merge_sort()) {
		cout << "Merge Sort" << endl;
		print_arr(init_seq);
	} else {
		cout << "Insertion Sort" << endl;
		insert_sort();
		print_arr(partially_sorted_seq);
	}

	return 0;
}